lib/evaluate/fold-integer.cpp

```diff
@@ -509,7 +509,15 @@ Expr<Type<TypeCategory::Integer, KIND>> FoldIntrinsicFunction(
           cx->u)};
     }
   } else if (name == "rank") {
-    // TODO assumed-rank dummy argument
+    if (const auto *array{UnwrapExpr<Expr<SomeType>>(args[0])}) {
+      if (auto named{ExtractNamedEntity(*array)}) {
+        const Symbol &symbol{named->GetLastSymbol()};
+        if (semantics::IsAssumedRankArray(symbol)) {
+          context.messages().Say("Must be a constant value"_err_en_US);
```

Comment:
Thank you @jeanPerier for review
Using `return Expr<T>{std::move(funcRef)};` works perfectly.
```

diff --git a/lib/evaluate/fold-integer.cpp b/lib/evaluate/fold-integer.cpp
index 6bf4f6c5..40ee110e 100644
--- a/lib/evaluate/fold-integer.cpp
+++ b/lib/evaluate/fold-integer.cpp
@@ -7,6 +7,7 @@
 //===----------------------------------------------------------------------===//
#include "fold-implementation.h"
+#include "flang/evaluate/variable.h"

 namespace Fortran::evaluate {

@@ -510,10 +511,10 @@ Expr<Type<TypeCategory::Integer, KIND>> FoldIntrinsicFunction(
     }
   } else if (name == "rank") {
     if (const auto *array{UnwrapExpr<Expr<SomeType>>(args[0])}) {
-      if (auto named{ExtractNamedEntity(*array)}) {
+      if (const auto named{ExtractNamedEntity(*array)}) {
         const Symbol &symbol{named->GetLastSymbol()};
         if (semantics::IsAssumedRankArray(symbol)) {
-          context.messages().Say("Must be a constant value"_err_en_US);
+          return Expr<T>{DescriptorInquiry{*named,DescriptorInquiry::Field::Rank}};
         }
         return Expr<T>{args[0].value().Rank()};
       }

`````

Not sure where does the following fails while compiling, but `DescriptorInquiry` expects first argument as a non-optional argument, whereas `named` is not.

Reply:
`DescriptorInquiry` cannot be wrapped in every kind of integer expressions. It can only be placed inside integer expressions of KIND `DescriptorInquiry::Result::kind` (see [here](https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/include/flang/evaluate/expression.h#L515))
The reason for this is to avoid implicitly narrowing descriptor fields (for lengths and dimensions, that could be problematic). Conversions have to be explicitly inserted in the expression tree.
So you would have to place the `DescriptorInquiry` into an `Expr<Type<TypeCategory::Integer, DescriptorInquiry::Result::kind>>` first  and then use `ConvertToType<T>` to convert this to the desired type.

Reply:
Note that you can use `ConvertToType<T>` on expression that are already of type T, this will simply return the original expression.

Reply:
I see `DescriptorInquiry`  [here](https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/include/flang/evaluate/expression.h#L517) placed inside `class Expr<Type<TypeCategory::Integer, KIND>>` in [file](https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/include/flang/evaluate/expression.h#L500)

If I am correct, what you expect is place the `DescriptorInquiry` into above class and specifically inside `using DescriptorInquiries` which is already present. 
I am calling `ConvertToType` in following manner inside `TODO`, which is causing an error.
`return Fold(context, ConvertToType<T>(std::move(*array)));`


Correct me if wrong, Do we need some implementation similar to `GetLowerBound` https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/lib/evaluate/fold-integer.cpp#L44
and use it in following way, where GetLowerBound replaces with implementation for extracting rank, with various checks in it.
`return Fold(context,
                ConvertToType<T>(GetLowerBound(context, *named, *dim)));`

Reply:
> I see `DescriptorInquiry` [here](https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/include/flang/evaluate/expression.h#L517) placed inside `class Expr<Type<TypeCategory::Integer, KIND>>` in [file](https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/include/flang/evaluate/expression.h#L500)

Let me explain a bit more what is happening  in the code your at pointing at ("here"). The specialization of `Epxr<T>` class for integers is defining a variant  member `u` which is the union of all possible operations an integer expression can be. This union type is built by combining the tuple types `Conversions `, `Operations`, `Indices`, `Indices`, and `DescriptorInquiries`. These intermediate types are just used for clarity and to help modifying what an integer expression can be depending on its kind.
 For `DescriptorInquiries`, the important point is to notice that it is defined conditionally using `std::conditional_t`. `DescriptorInquires` will be an empty tuple for all integer KIND but one ( for `DescriptorInquiry::Result::kind`, currently equal to 8). In practice, this means that `Expr<Type<TypeCategory::Integer, KIND>>` cannot directly hold a `DescriptorInquiry`, unless KIND is equal to `DescriptorInquiry::Result::kind` (currently equal to 8).

You could read this definition in a BNF form:

_some-integer-expression_  **is**  _integer-8-expression_ **or** _other-integer-expression_
_integer-8-expression_ **is** _descriptor-inquiry_ **or**  _other-integer-sub-expressions_
_other-integer-expression_ **is** _other-integer-sub-expressions_
_other-integer-sub-expressions_ **is** _integer-function-call_ **or** ...

> If I am correct, what you expect is place the `DescriptorInquiry` into above class and specifically inside `using DescriptorInquiries` which is already present.

Inside fold-integer.h code, `Expr<T>` may be an _other-integer-expression_ from my BNF example  above. So, what I expect you is to always place the newly built `DescriptorInquiry` into an _integer-8-expression_  first (build an `Expr<Type<TypeCategory::Integer, DescriptorInquiry::Result::kind>>{`DescriptorInquiry{...}}`) and then convert this into the actual integer type that the expression should be folded to using `ConvertToType<T>`.

> I am calling `ConvertToType` in following manner inside `TODO`, which is causing an error.
> `return Fold(context, ConvertToType<T>(std::move(*array)));`

`std::move(*array` is not what you want to convert, you need to convert the newly built "_integer-8-expression_ ".

> Correct me if wrong, Do we need some implementation similar to `GetLowerBound`
> 
> https://github.com/flang-compiler/f18/blob/35fd0cda58776389d2ed68eaefbc1e1d59423ec8/lib/evaluate/fold-integer.cpp#L44
> 
> 
> and use it in following way, where GetLowerBound replaces with implementation for extracting rank, with various checks in it.
> `return Fold(context, ConvertToType<T>(GetLowerBound(context, *named, *dim)));`

It is very similar. You could move the RANK folding implementation part that is inside `if (auto named{ExtractNamedEntity(*array)}) {` into a `GetRank` function that would return an expression. In my opinion, the RANK folding code is pretty small and there is currently no need elsewhere for  `GetRank` that I know. So it's also OK if you directly implement everything in RANK folding code.

Reply:
Thanks for detailed explanation.